google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -352,6 +351,328 @@ public void sendWritesToDifferentDocsInSameBatch() throws Exception {
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(1));
+          }
+        };
+    responseStubber.initializeStub(batchWriteCapture, firestoreMock);
+
+    bulkWriter.addWriteErrorListener(
+        new WriteErrorCallback() {
+          public boolean onError(BulkWriterError error) {
+            return true;
+          }
+        });
+
+    ApiFuture<WriteResult> result1 = bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close();
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(1, 0), result1.get().getUpdateTime());
+  }
+
+  @Test
+  public void retriesMaintainCorrectWriteResolutionOrdering() throws Exception {
+    ResponseStubber responseStubber =
+        new ResponseStubber() {
+          {
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                failedResponse(Code.INTERNAL_VALUE));
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(1));
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc2")),
+                successResponse(2));
+          }
+        };
+    responseStubber.initializeStub(batchWriteCapture, firestoreMock);
+    ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+
+    // Use separate futures to track listener completion since the callbacks are run on a different
+    // thread than the BulkWriter operations.
+    final SettableApiFuture<Void> flushComplete = SettableApiFuture.create();
+
+    final List<String> operations = new ArrayList<>();
+    bulkWriter.addWriteErrorListener(
+        new WriteErrorCallback() {
+          public boolean onError(BulkWriterError error) {
+            return true;
+          }
+        });
+
+    bulkWriter
+        .set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP)
+        .addListener(
+            new Runnable() {
+              public void run() {
+                operations.add("BEFORE_FLUSH");
+              }
+            },
+            executor);
+    bulkWriter
+        .flush()
+        .addListener(
+            new Runnable() {
+              public void run() {
+                operations.add("FLUSH");
+                flushComplete.set(null);
+              }
+            },
+            executor);
+    bulkWriter
+        .set(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP)
+        .addListener(
+            new Runnable() {
+              public void run() {
+                operations.add("AFTER_FLUSH");
+              }
+            },
+            executor);
+    flushComplete.get();
+
+    // Verify that the 2nd operation did not complete as a result of the flush() call.
+    assertArrayEquals(new String[] {"BEFORE_FLUSH", "FLUSH"}, operations.toArray());
+    bulkWriter.close();
```

Comment:
It might make sense to add another flush here and verify "AFTER_FLUSH" as well. That way, it doesn't look unused.